### PR TITLE
Fix dynamically excluded providers caching

### LINKED ISSUE #3380: Dynamically filtered providers cache is set incorrectly
## Description

Dynamically filtered-out providers (the content providers that have "Hide content" set to `True` in the Django Admin) are cached with the cache key `:1:filtered_providers`, but the value that is saved there is incorrect.

@AetherUnbound, I learned that when we use Django's `cache` here (and not `django_redis` directly as in many other places), Django transforms the key, adding a blank `key_prefix` and `1` as a version, so `filtered_providers` becomes `:1:filtered_providers` [^1].

[^1]: [Cache key transformation](https://docs.djangoproject.com/en/4.2/topics/cache/#cache-key-transformation)

## Screenshots

Redis value for `:1:filtered_providers` key:

```
{"_db":null,"_hints":{},"_query":{"alias_refcount":{"content_provider":2},"alias_map":{"content_provider":{"table_name":"content_provider","table_alias":"content_provider"}},"alias_cols":true,"external_aliases":{},"table_map":{"content_provider":["content_provider"]},"used_aliases":[],"where":{"children":[{"_constructor_args":[[{"_constructor_args":[["content_provider",{}],{}],"output_field":{},"alias":"content_provider","target":{},"contains_aggregate":false,"contains_over_clause":false},true],{}],"lhs":{"_constructor_args":[["content_provider",{}],{}],"output_field":{},"alias":"content_provider","target":{},"contains_aggregate":false,"contains_over_clause":false},"rhs":true,"bilateral_transforms":[],"contains_aggregate":false,"contains_over_clause":false}],"connector":"AND","negated":false,"contains_aggregate":false,"contains_over_clause":false},"annotations":{},"extra":{},"_filtered_relations":{},"_annotation_select_cache":null,"filter_is_sticky":false,"_lookup_joins":["content_provider"],"select_related":false,"deferred_loading":[[],true],"select":[{"_constructor_args":[["content_provider",{}],{}],"output_field":{},"alias":"content_provider","target":{}}],"values_select":["provider_identifier"],"has_select_fields":true,"extra_select_mask":[],"_extra_select_cache":null,"annotation_select_mask":[],"base_table":"content_provider","default_cols":false},"_result_cache":[{"provider_identifier":"floraon"}],"_sticky_filter":false,"_for_write":false,"_prefetch_related_lookups":[],"_prefetch_done":false,"_known_related_objects":{},"_fields":["provider_identifier"],"_defer_next_filter":false,"_deferred_filter":null,"_django_version":"4.2.7"}
```

### PR DESCRIPTION
**Pull request by obulat**
## Fixes

Fixes #3380 by @obulat

## Description

This PR saves a list of `provider_identifier`s for the dynamically-excluded providers to the redis cache, instead of saving the QuerySet.
It also adds two unit tests for the specific functions, but there are also several other tests that test dynamically excluded providers using both cache and the `ContentProvider` model saved in the database.

## Testing Instructions

Run the app, go to `http://localhost:50280/admin`, select one of the `ContentProvider` models and check its "Hide Content" property.
Screenshot 2023-11-21 at 5 22 01 PM

 Execute a search (http://localhost:50280/v1/images). Check Redis value for `:1:filtered_providers` (I used the RedisInsights GUI): 

Screenshot 2023-11-21 at 5 23 35 PM

Also, check that on `main`, if you set "Hide content" to true for one of the providers, the value saved in Redis should look like the one in the related issue instead of a list of filtered providers.

## Checklist

- [ ] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [ ] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: api/api/controllers/search_controller.py [hunk 1]
```diff
@@ -170,19 +171,23 @@ def get_excluded_providers_query() -> Q | None:
...
-    filtered_providers = cache.get(key=filter_cache_key)
+    filtered_providers = cache.get(key=FILTERED_PROVIDERS_CACHE_KEY)
     if not filtered_providers:
-        filtered_providers = models.ContentProvider.objects.filter(
-            filter_content=True
-        ).values("provider_identifier")
+        providers = models.ContentProvider.objects.filter(filter_content=True).values(
+            "provider_identifier"
+        )
+        filtered_providers = [p["provider_identifier"] for p in providers]
```
Comment (sarayourfriend) [inline api/api/controllers/search_controller.py:183]: nit: Django has `values_list` that with `flat` that can simplify this (remove the intermediary dict):

```suggestion
        filtered_providers = list(models.ContentProvider.objects.filter(filter_content=True).values_list(
            "provider_identifier",
            flat=True,
        ))
```

**Review by sarayourfriend — APPROVED** [review]
LGTM!

File: api/api/controllers/search_controller.py (hunk 1 above)
Comment (obulat) [reply api/api/controllers/search_controller.py:183]: TIL, thank you!

File: api/api/controllers/search_controller.py [hunk 2]
```diff
@@ -170,19 +171,23 @@ def get_excluded_providers_query() -> Q | None:
...
-    filtered_providers = cache.get(key=filter_cache_key)
+    filtered_providers = cache.get(key=FILTERED_PROVIDERS_CACHE_KEY)
     if not filtered_providers:
-        filtered_providers = models.ContentProvider.objects.filter(
-            filter_content=True
-        ).values("provider_identifier")
+        filtered_providers = list(models.ContentProvider.objects.filter(filter_content=True).values_list(
+            "provider_identifier",
+            flat=True,
+        ))
```
Comment (krysal) [inline api/api/controllers/search_controller.py:183]: This should fix the linting:
```suggestion
        filtered_providers = list(
            models.ContentProvider.objects.filter(filter_content=True).values_list(
                "provider_identifier", flat=True
            )
        )
```

**Review by krysal — APPROVED** [review]
Well seen! LGTM!

**Pull request merged**